### PR TITLE
Feb 21, 2024 Spark Spell

### PR DESCRIPTION
**Pull request by hexonaut**
* Increase DAI APY to 6.7%
 * Increase wstETH supply cap to 1.2m

Poll/forum posts pending.

### CODE REVIEW

File: src/proposals/20240221/SparkEthereum_20240221.sol [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+contract SparkEthereum_20240221 is SparkPayloadEthereum {
+
+    address public constant DAI    = 0x6B175474E89094C44Da98b954EedeAC495271d0F;
+    address public constant WSTETH = 0x7f39C581F595B53c5cb19bD0b3f8dA6c935E2Ca0;
+
+    // Formula for 6.7% target APY:
+    // bc -l <<< 'scale=27; (e( l(1.067)/(60 * 60 * 24 * 365) ) - 1) * 60 * 60 * 24 * 365 - 0.048790164207174267760128000'
+    address public constant DAI_IRM = 0x3C4B090b5b479402e2270C66461D6a62B2054198;
+
+
```
Comment (barrutko) [inline src/proposals/20240221/SparkEthereum_20240221.sol:23]: Redundant line

File: src/proposals/20240221/SparkEthereum_20240221.t.sol [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+interface IRateSource {
+    function getAPR() external view returns (int256);
+}
+
+contract SparkEthereum_20240221Test is SparkEthereumTestBase {
+
+    address public constant OLD_DAI_INTEREST_RATE_STRATEGY  = 0x512AFEDCF6696d9707dCFECD4bdc73e9902e3c6A;
+    address public constant NEW_DAI_INTEREST_RATE_STRATEGY  = 0x3C4B090b5b479402e2270C66461D6a62B2054198;
+
+    int256 public constant DAI_IRM_SPREAD =  0.016060808179122167684448000e27;
```
Comment (barrutko) [inline src/proposals/20240221/SparkEthereum_20240221.t.sol:19]:
```suggestion
    int256 public constant DAI_IRM_SPREAD = 0.016060808179122167684448000e27;
```

File: src/proposals/20240221/SparkEthereum_20240221.t.sol [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+    uint256 public constant WSTETH_SUPPLY_CAP_OLD = 800_000;
+    uint256 public constant WSTETH_SUPPLY_CAP_NEW = 1_200_000;
```
Comment (barrutko) [inline src/proposals/20240221/SparkEthereum_20240221.t.sol:22]: I'd rename it this way, so it follows the convention
```suggestion
    uint256 public constant OLD_WSTETH_SUPPLY_CAP = 800_000;
    uint256 public constant NEW_WSTETH_SUPPLY_CAP = 1_200_000;
```

File: src/proposals/20240221/SparkEthereum_20240221.t.sol [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+interface IRateSource {
+    function getAPR() external view returns (int256);
+}
+
+contract SparkEthereum_20240221Test is SparkEthereumTestBase {
+
+    address public constant OLD_DAI_INTEREST_RATE_STRATEGY  = 0x512AFEDCF6696d9707dCFECD4bdc73e9902e3c6A;
+    address public constant NEW_DAI_INTEREST_RATE_STRATEGY  = 0x3C4B090b5b479402e2270C66461D6a62B2054198;
```
Comment (barrutko) [inline src/proposals/20240221/SparkEthereum_20240221.t.sol:17]:
```suggestion
    address public constant OLD_DAI_INTEREST_RATE_STRATEGY = 0x512AFEDCF6696d9707dCFECD4bdc73e9902e3c6A;
    address public constant NEW_DAI_INTEREST_RATE_STRATEGY = 0x3C4B090b5b479402e2270C66461D6a62B2054198;
```

File: src/proposals/20240221/SparkEthereum_20240221.t.sol [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        ReserveConfig memory wstethConfigBefore = _findReserveConfigBySymbol(allConfigsBefore, 'wstETH');
+
+        assertEq(daiConfigBefore.interestRateStrategy, OLD_DAI_INTEREST_RATE_STRATEGY);
+        assertEq(wstethConfigBefore.supplyCap, WSTETH_SUPPLY_CAP_OLD);
+
+        GovHelpers.executePayload(vm, payload, executor);
+
+        ReserveConfig[] memory allConfigsAfter = createConfigurationSnapshot('', pool);
+
+        ReserveConfig memory daiConfigAfter     = _findReserveConfigBySymbol(allConfigsAfter, 'DAI');
```
Comment (barrutko) [inline src/proposals/20240221/SparkEthereum_20240221.t.sol:48]:
```suggestion
        ReserveConfig memory daiConfigAfter = _findReserveConfigBySymbol(allConfigsAfter, 'DAI');
```

File: src/proposals/20240221/SparkEthereum_20240221.t.sol [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+        int256 potDsrApr = IRateSource(IIRM(daiConfigAfter.interestRateStrategy).RATE_SOURCE()).getAPR();
```
Comment (barrutko) [inline src/proposals/20240221/SparkEthereum_20240221.t.sol:50]: Before that, I'd also check that `IIRM(daiConfigAfter.interestRateStrategy).RATE_SOURCE()` is equal to a proper hardcoded address

Comment (barrutko) [reply src/proposals/20240221/SparkEthereum_20240221.t.sol:50]: Checked on etherscan, it's ok, but I'd still test it just for clarity

**Review by barrutko — CHANGES_REQUESTED** [review]
Logic and numbers look good.
Some style changes and one additional test requested.
Vote links still to be added.

File: src/proposals/20240221/SparkEthereum_20240221.sol (hunk 1 above)
Comment (hexonaut) [reply src/proposals/20240221/SparkEthereum_20240221.sol:23]: fixed

File: src/proposals/20240221/SparkEthereum_20240221.t.sol (hunk 2 above)
Comment (hexonaut) [reply src/proposals/20240221/SparkEthereum_20240221.t.sol:19]: fixed

File: src/proposals/20240221/SparkEthereum_20240221.t.sol (hunk 3 above)
Comment (hexonaut) [reply src/proposals/20240221/SparkEthereum_20240221.t.sol:22]: fixed

File: src/proposals/20240221/SparkEthereum_20240221.t.sol (hunk 4 above)
Comment (hexonaut) [reply src/proposals/20240221/SparkEthereum_20240221.t.sol:17]: fixed

File: src/proposals/20240221/SparkEthereum_20240221.t.sol (hunk 5 above)
Comment (hexonaut) [reply src/proposals/20240221/SparkEthereum_20240221.t.sol:48]: fixed

File: src/proposals/20240221/SparkEthereum_20240221.t.sol (hunk 6 above)
Comment (hexonaut) [reply src/proposals/20240221/SparkEthereum_20240221.t.sol:50]: I check that it's the same as before. That seems sufficient to me, but lmk if you want hardcoded address.

Comment (barrutko) [reply src/proposals/20240221/SparkEthereum_20240221.t.sol]: Ah yes, I missed it. Yeah, I think it's sufficient.

**Review by barrutko — COMMENTED** [review]
LGTM
I'll wait with the approval for the vote links though

**Review by barrutko — COMMENTED** [review]
New changes look good.
Still waiting for the poll link/s.

**Review by lucas-manuel — COMMENTED** [review]
Checked:

- Links
- Addresses in spell
- Supply cap
- Deployment params in IRM
- Verified resulting APR from base rate + spread equals desired 6.7% APY with this formula: 
```
bc -l <<< ‘scale=27; e(0.064850972386296435444576000) - 1’
.067000000071147689450689538
```
- Checked diff
- Reviewed tests
- Can't get bytecode verification to work with sourcify yet, but otherwise LGTM

**Review by lucas-manuel — DISMISSED** [review]
Verified bytecode here: https://github.com/marsfoundation/bytecode-verification/blob/faf852201fd706204c97ca277afa8a065a39d0cc/test/Bytecode.t.sol#L64

LGTM

**Review by barrutko — APPROVED** [review]
LGTM

**Review by lucas-manuel — APPROVED** [review]

**Pull request merged**